src/tactic/congrm.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import tactic.interactive
+
+/-!  `congrm`: `congr` with pattern-matching -/
+
+namespace tactic.interactive
+open tactic interactive
+setup_tactic_parser
+
+/--  Scans three `expr`s `e, lhs, rhs` in parallel.
+Currently, `equate_with_pattern` has three behaviours at each location:
+produce a goal, recurse, or skip.
+
+See the doc-string for `tactic.interactive.dap` for more details.
+-/
+meta def equate_with_pattern : expr → expr → expr → tactic unit
+| (expr.app f e) (expr.app f0 e0) (expr.app f1 e1) := do
+  match e with
+  | (expr.mvar _ _ _) := do
+    el ← mk_app `eq [e0, e1],
+    n ← get_unused_name "h",
+    assert n el,
+    rotate,
+    get_local n >>= rewrite_target,
```

Comment:
A key feature of the `congr` tactic is that it uses the automatically generated congruence lemmas (`mk_specialized_congr_lemma`).  This is what allows it to ignore subsingleton arguments and rewrite in dependent positions.  For example on `ite p a b = ite q a b` we want the subgoal `p = q` but we cannot `rw` with `p = q`.  Similarly on `ite p a b = by classical; exact p a c` we want the subgoal `b = c` but this requires knowing that `decidable p` is a subsingleton.

In other words, `equate_with_pattern` should do something like `apply_eq_congr_core`, except compute the congruence lemma from the pattern instead of the lhs.

It would also be cool if congr(m) could descend into lambdas and foralls.  I wonder why we never implemented that.

Reply:
> It would also be cool if congr(m) could descend into lambdas and foralls. I wonder why we never implemented that.

This is my biggest gripe with `congr`.